### PR TITLE
added product addon api for post/put/delete

### PR DESCRIPTION
**Pull request by stemod**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ CHANGELOG for Sulu Product Bundle
 * feature/addons
     * ENHANCEMENT Added GET-api for product addons
     * ENHANCEMENT Added GET-api for currencies
+    * ENHANCEMENT Added POST, PUT, DELETE api for product addons
```
Comment (turbo-ele) [inline CHANGELOG.md]: please add to first line

File: Controller/AddonController.php [hunk 2]
```diff
@@ -68,6 +71,66 @@ public function getProductAddonsAction(Request $request, $productId)
...
+        $locale = $this->getProductLocaleManager()->retrieveLocale($this->getUser(), $request->get('locale'));
+
+        $addonId = $request->get('addon');
+        $prices = $request->get('prices');
+
+        $addon = $this->getManager()->saveProductAddon($productId, $addonId, $locale, $prices);
+
+        $view = $this->view($addon, 200);
+
+        return $this->handleView($view);
```
Comment (turbo-ele) [inline Controller/AddonController.php]: post and put seem to do the exact same thing

File: Controller/AddonController.php [hunk 3]
```diff
@@ -68,6 +71,66 @@ public function getProductAddonsAction(Request $request, $productId)
...
+    }
+
+    /**
+     * @Delete("/products/{productId}/addons/{addonId}")
+     *
+     * @param $productId
+     * @param $addonId
```
Comment (turbo-ele) [inline Controller/AddonController.php]: add types

File: Controller/AddonController.php [hunk 4]
```diff
@@ -68,6 +71,66 @@ public function getProductAddonsAction(Request $request, $productId)
...
+     *
+     * @return Response
+     */
+    public function deleteProductAddonAction($productId, $addonId)
+    {
+        $this->getManager()->deleteProductAddon($productId, $addonId);
+        $view = $this->view();
```
Comment (turbo-ele) [inline Controller/AddonController.php]: add http status code 204

File: Docs/product-addon.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            "id": 1,
+            "addon": {
+                "id": 4,
+                "name": "Addon 1",
+                "type": {
+                    "id": 3,
+                    "name": "Produkt Erweiterung"
+                },
+
+                ...
```
Comment (turbo-ele) [inline Docs/product-addon.md]: remove

File: Docs/product-addon.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                    "currency": {
+                        "name": "Euro",
+                        "id": 2,
+                        "code": "EUR",
+                        "number": "978"
+                    },
+                    "price": "222"
+                }
+            ]
+        ]
```
Comment (turbo-ele) [inline Docs/product-addon.md]: add `}` before this line

File: Product/ProductAddonManager.php [hunk 7]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
 
         return $apiAddons;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function saveProductAddon($productId, $addonId, $locale, array $prices = null)
+    {
+        $productAddon = $this->addonRepository->findOneBy(['product' => $productId, 'addon' => $addonId]);
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: i.m.o. you should break associative array into multiple lines.

File: Product/ProductAddonManager.php [hunk 8]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+        if (!$productAddon) {
+            $product = $this->productRepository->findById($productId);
+            $addon = $this->productRepository->findById($addonId);
+
+            if (!$product) {
+                throw new EntityNotFoundException('SuluProductBundle:Product', $productId);
+            }
+            if (!$addon) {
+                throw new EntityNotFoundException('SuluProductBundle:Product', $addonId);
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: SuluProductBundle:Addon

File: Product/ProductAddonManager.php [hunk 9]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+            }
+            $addonPrice->setCurrency($currency);
+
+            $addonPrices[] = $addonPrice;
+        }
+
+        /** @var AddonPrice $savedPrice */
+        // Delete old entries
+        foreach ($productAddon->getAddonPrices()->toArray() as $savedPrice) {
+            $found = false;
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: $isFound

File: Product/ProductAddonManager.php [hunk 10]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+            foreach ($addonPrices as $addonPrice) {
+                if ($addonPrice->getCurrency() === $savedPrice->getCurrency()) {
+                    $found = true;
+                }
+            }
+            if (!$found) {
+                $this->entityManager->remove($savedPrice);
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: add `$addonPrice->removePrice($savedPrice)` ?

Comment (turbo-ele) [inline Product/ProductAddonManager.php]: move this line directly into if of foreach and add break

File: Product/ProductAddonManager.php [hunk 11]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+        }
+
+        /** @var AddonPrice $addonPrice */
+        // Save/Update entries
+        foreach ($addonPrices as $addonPrice) {
+            $new = true;
+            foreach ($productAddon->getAddonPrices()->toArray() as $savedPrice) {
+                if ($addonPrice->getCurrency() === $savedPrice->getCurrency()) {
+                    $savedPrice->setPrice($addonPrice->getPrice());
+                    $new = false;
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: add break

File: Product/ProductAddonManager.php [hunk 12]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+            }
+            if (!$found) {
+                $this->entityManager->remove($savedPrice);
+            }
+        }
+
+        /** @var AddonPrice $addonPrice */
+        // Save/Update entries
+        foreach ($addonPrices as $addonPrice) {
+            $new = true;
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: $isNew

File: Product/ProductAddonManager.php [hunk 13]
```diff
@@ -45,9 +76,95 @@ public function findAddonsByProductIdAndLocale($id, $locale)
...
+                    $new = false;
+                }
+            }
+            if ($new) {
+                $this->entityManager->persist($addonPrice);
+                $productAddon->addAddonPrice($addonPrice);
+            }
+        }
+
+        $this->entityManager->flush();
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: do we need to flush here?

File: Product/ProductAddonManagerInterface.php [hunk 14]
```diff
@@ -21,4 +22,20 @@
      * @return Addon[]
      */
     public function findAddonsByProductIdAndLocale($id, $locale);
+
+    /**
+     * @param int $product
+     * @param int $addon
+     * @param string $locale
+     * @param array $prices
```
Comment (turbo-ele) [inline Product/ProductAddonManagerInterface.php]: `@param AddonPrice[] $prices`

File: Controller/AddonController.php (hunk 4 above)
Comment (stemod) [inline Controller/AddonController.php]: ... is added automatically.

File: Product/ProductAddonManager.php (hunk 8 above)
Comment (stemod) [inline Product/ProductAddonManager.php]: no $addon is a Product

File: Product/ProductAddonManager.php (hunk 13 above)
Comment (stemod) [inline Product/ProductAddonManager.php]: yes

File: Product/ProductAddonManagerInterface.php (hunk 14 above)
Comment (stemod) [inline Product/ProductAddonManagerInterface.php]: array is correct - these are note AddonPrices

File: Controller/AddonController.php (hunk 4 above)
Comment (turbo-ele) [inline Controller/AddonController.php]: oh, ok ;)

File: Product/ProductAddonManager.php (hunk 8 above)
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: right

File: Product/ProductAddonManagerInterface.php (hunk 14 above)
Comment (turbo-ele) [inline Product/ProductAddonManagerInterface.php]: ok

File: Product/ProductAddonManager.php [hunk 15]
```diff
@@ -85,6 +85,17 @@ public function findAddonsByProductIdAndLocale($id, $locale)
     /**
      * {@inheritdoc}
      */
+    public function findAddonById($id, $locale)
+    {
+        /** @var Addon $addon */
+        $addon = $this->addonRepository->find($id);
+
+        return new ApiAddon($addon, $this->productFactory, $locale);
```
Comment (turbo-ele) [inline Product/ProductAddonManager.php]: what if addon cannot be found?

**Comment by turbo-ele** [comment]
else -when travis tests have been successful - ack

**Pull request merged**